File: python/InOrderBinaryTreeTraversal.py
# https://leetcode.com/problems/binary-tree-inorder-traversal/

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        """
        Given a binary tree, return the inorder traversal of its nodes' values.
        """

        nodes = []
        visited_vals = []

        curr = root

        while curr is not None or len(nodes) > 0:
            # reach leftmost node of curr
            while curr is not None:
                nodes.append(curr)
                curr = curr.left

            # curr is None
            curr = nodes.pop()
            visited_vals.append(curr.val)

            # visit right node
            curr = curr.right

        return visited_vals

File: java/InOrderBinaryTreeTraversal.java
package com.leetcode.medium;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

class TreeNode {
    TreeNode left;
    TreeNode right;
    int val;
}

public class InOrderBinaryTreeTraversal {
    
    public List<Integer> list = new ArrayList<>();

    public List<Integer> inorderTraversal(TreeNode root) {
        if (root == null) {
            return list;
        }

        Stack<TreeNode> stack = new Stack<>();
        TreeNode current = root;
        // stack.push(current);

        while (current != null || !stack.empty()) {
            while (current != null) {
                stack.push(current);
                current = current.left;
            }

            current = stack.pop();

            list.add(current.val);
            current = current.right;
        }

        return list;
    }
}
